### PR TITLE
Add bit-precise overloads for builtin operators 

### LINKED ISSUE #82998: [clang] _BitInt member of struct/class implicitly deletes operator==
https://godbolt.org/z/134fjse57
```cpp
struct A
{
    _BitInt(64) a;

    constexpr bool operator==(A const& a) const = default; // implicitly deleted due to _BitInt(64)
};

bool cmp(A a, A b)
{
    return a == b; // error
}
```
```
<source>:5:20: warning: explicitly defaulted equality comparison operator is implicitly deleted [-Wdefaulted-function-deleted]
    5 |     constexpr bool operator==(A const& a) const = default; // implicitly deleted due to _BitInt(64)
      |                    ^
<source>:3:17: note: defaulted 'operator==' is implicitly deleted because implied comparison for member 'a' is ambiguous
    3 |     _BitInt(64) a;
      |                 ^
note: built-in candidate operator(float, float)
note: built-in candidate operator(float, double)
note: built-in candidate operator(float, long double)
note: built-in candidate operator(float, __float128)
note: built-in candidate operator(float, int)
note: built-in candidate operator(float, long)
note: built-in candidate operator(float, long long)
note: built-in candidate operator(float, __int128)
note: built-in candidate operator(float, unsigned int)
note: built-in candidate operator(float, unsigned long)
note: built-in candidate operator(float, unsigned long long)
note: built-in candidate operator(float, unsigned __int128)
note: built-in candidate operator(double, float)
note: built-in candidate operator(double, double)
note: built-in candidate operator(double, long double)
note: built-in candidate operator(double, __float128)
note: built-in candidate operator(double, int)
note: built-in candidate operator(double, long)
note: built-in candidate operator(double, long long)
note: built-in candidate operator(double, __int128)
note: built-in candidate operator(double, unsigned int)
note: built-in candidate operator(double, unsigned long)
note: built-in candidate operator(double, unsigned long long)
note: built-in candidate operator(double, unsigned __int128)
note: built-in candidate operator(long double, float)
note: built-in candidate operator(long double, double)
note: built-in candidate operator(long double, long double)
note: built-in candidate operator(long double, __float128)
note: built-in candidate operator(long double, int)
note: built-in candidate operator(long double, long)
note: built-in candidate operator(long double, long long)
note: built-in candidate operator(long double, __int128)
note: built-in candidate operator(long double, unsigned int)
note: built-in candidate operator(long double, unsigned long)
note: built-in candidate operator(long double, unsigned long long)
note: built-in candidate operator(long double, unsigned __int128)
note: built-in candidate operator(__float128, float)
note: built-in candidate operator(__float128, double)
note: built-in candidate operator(__float128, long double)
note: built-in candidate operator(__float128, __float128)
note: built-in candidate operator(__float128, int)
note: built-in candidate operator(__float128, long)
note: built-in candidate operator(__float128, long long)
note: built-in candidate operator(__float128, __int128)
note: built-in candidate operator(__float128, unsigned int)
note: built-in candidate operator(__float128, unsigned long)
note: built-in candidate operator(__float128, unsigned long long)
note: built-in candidate operator(__float128, unsigned __int128)
note: built-in candidate operator(int, float)
note: built-in candidate operator(int, double)
note: built-in candidate operator(int, long double)
note: built-in candidate operator(int, __float128)
note: built-in candidate operator(int, int)
note: built-in candidate operator(int, long)
note: built-in candidate operator(int, long long)
note: built-in candidate operator(int, __int128)
note: built-in candidate operator(int, unsigned int)
note: built-in candidate operator(int, unsigned long)
note: built-in candidate operator(int, unsigned long long)
note: built-in candidate operator(int, unsigned __int128)
note: built-in candidate operator(long, float)
note: built-in candidate operator(long, double)
note: built-in candidate operator(long, long double)
note: built-in candidate operator(long, __float128)
note: built-in candidate operator(long, int)
note: built-in candidate operator(long, long)
note: built-in candidate operator(long, long long)
note: built-in candidate operator(long, __int128)
note: built-in candidate operator(long, unsigned int)
note: built-in candidate operator(long, unsigned long)
note: built-in candidate operator(long, unsigned long long)
note: built-in candidate operator(long, unsigned __int128)
note: built-in candidate operator(long long, float)
note: built-in candidate operator(long long, double)
note: built-in candidate operator(long long, long double)
note: built-in candidate operator(long long, __float128)
note: built-in candidate operator(long long, int)
note: built-in candidate operator(long long, long)
note: built-in candidate operator(long long, long long)
note: built-in candidate operator(long long, __int128)
note: built-in candidate operator(long long, unsigned int)
note: built-in candidate operator(long long, unsigned long)
note: built-in candidate operator(long long, unsigned long long)
note: built-in candidate operator(long long, unsigned __int128)
note: built-in candidate operator(__int128, float)
note: built-in candidate operator(__int128, double)
note: built-in candidate operator(__int128, long double)
note: built-in candidate operator(__int128, __float128)
note: built-in candidate operator(__int128, int)
note: built-in candidate operator(__int128, long)
note: built-in candidate operator(__int128, long long)
note: built-in candidate operator(__int128, __int128)
note: built-in candidate operator(__int128, unsigned int)
note: built-in candidate operator(__int128, unsigned long)
note: built-in candidate operator(__int128, unsigned long long)
note: built-in candidate operator(__int128, unsigned __int128)
note: built-in candidate operator(unsigned int, float)
note: built-in candidate operator(unsigned int, double)
note: built-in candidate operator(unsigned int, long double)
note: built-in candidate operator(unsigned int, __float128)
note: built-in candidate operator(unsigned int, int)
note: built-in candidate operator(unsigned int, long)
note: built-in candidate operator(unsigned int, long long)
note: built-in candidate operator(unsigned int, __int128)
note: built-in candidate operator(unsigned int, unsigned int)
note: built-in candidate operator(unsigned int, unsigned long)
note: built-in candidate operator(unsigned int, unsigned long long)
note: built-in candidate operator(unsigned int, unsigned __int128)
note: built-in candidate operator(unsigned long, float)
note: built-in candidate operator(unsigned long, double)
note: built-in candidate operator(unsigned long, long double)
note: built-in candidate operator(unsigned long, __float128)
note: built-in candidate operator(unsigned long, int)
note: built-in candidate operator(unsigned long, long)
note: built-in candidate operator(unsigned long, long long)
note: built-in candidate operator(unsigned long, __int128)
note: built-in candidate operator(unsigned long, unsigned int)
note: built-in candidate operator(unsigned long, unsigned long)
note: built-in candidate operator(unsigned long, unsigned long long)
note: built-in candidate operator(unsigned long, unsigned __int128)
note: built-in candidate operator(unsigned long long, float)
note: built-in candidate operator(unsigned long long, double)
note: built-in candidate operator(unsigned long long, long double)
note: built-in candidate operator(unsigned long long, __float128)
note: built-in candidate operator(unsigned long long, int)
note: built-in candidate operator(unsigned long long, long)
note: built-in candidate operator(unsigned long long, long long)
note: built-in candidate operator(unsigned long long, __int128)
note: built-in candidate operator(unsigned long long, unsigned int)
note: built-in candidate operator(unsigned long long, unsigned long)
note: built-in candidate operator(unsigned long long, unsigned long long)
note: built-in candidate operator(unsigned long long, unsigned __int128)
note: built-in candidate operator(unsigned __int128, float)
note: built-in candidate operator(unsigned __int128, double)
note: built-in candidate operator(unsigned __int128, long double)
note: built-in candidate operator(unsigned __int128, __float128)
note: built-in candidate operator(unsigned __int128, int)
note: built-in candidate operator(unsigned __int128, long)
note: built-in candidate operator(unsigned __int128, long long)
note: built-in candidate operator(unsigned __int128, __int128)
note: built-in candidate operator(unsigned __int128, unsigned int)
note: built-in candidate operator(unsigned __int128, unsigned long)
note: built-in candidate operator(unsigned __int128, unsigned long long)
note: built-in candidate operator(unsigned __int128, unsigned __int128)
<source>:5:51: note: replace 'default' with 'delete'
    5 |     constexpr bool operator==(A const& a) const = default; // implicitly deleted due to _BitInt(64)
      |                                                   ^~~~~~~
      |                                                   delete
<source>:10:14: error: object of type 'A' cannot be compared because its 'operator==' is implicitly deleted
   10 |     return a == b; // error
      |              ^
<source>:5:20: note: explicitly defaulted function was implicitly deleted here
    5 |     constexpr bool operator==(A const& a) const = default; // implicitly deleted due to _BitInt(64)
      |                    ^
<source>:3:17: note: defaulted 'operator==' is implicitly deleted because implied comparison for member 'a' is ambiguous
    3 |     _BitInt(64) a;
      |                 ^
note: built-in candidate operator(float, float)
note: built-in candidate operator(float, double)
note: built-in candidate operator(float, long double)
note: built-in candidate operator(float, __float128)
note: built-in candidate operator(float, int)
note: built-in candidate operator(float, long)
note: built-in candidate operator(float, long long)
note: built-in candidate operator(float, __int128)
note: built-in candidate operator(float, unsigned int)
note: built-in candidate operator(float, unsigned long)
note: built-in candidate operator(float, unsigned long long)
note: built-in candidate operator(float, unsigned __int128)
note: built-in candidate operator(double, float)
note: built-in candidate operator(double, double)
note: built-in candidate operator(double, long double)
note: built-in candidate operator(double, __float128)
note: built-in candidate operator(double, int)
note: built-in candidate operator(double, long)
note: built-in candidate operator(double, long long)
note: built-in candidate operator(double, __int128)
note: built-in candidate operator(double, unsigned int)
note: built-in candidate operator(double, unsigned long)
note: built-in candidate operator(double, unsigned long long)
note: built-in candidate operator(double, unsigned __int128)
note: built-in candidate operator(long double, float)
note: built-in candidate operator(long double, double)
note: built-in candidate operator(long double, long double)
note: built-in candidate operator(long double, __float128)
note: built-in candidate operator(long double, int)
note: built-in candidate operator(long double, long)
note: built-in candidate operator(long double, long long)
note: built-in candidate operator(long double, __int128)
note: built-in candidate operator(long double, unsigned int)
note: built-in candidate operator(long double, unsigned long)
note: built-in candidate operator(long double, unsigned long long)
note: built-in candidate operator(long double, unsigned __int128)
note: built-in candidate operator(__float128, float)
note: built-in candidate operator(__float128, double)
note: built-in candidate operator(__float128, long double)
note: built-in candidate operator(__float128, __float128)
note: built-in candidate operator(__float128, int)
note: built-in candidate operator(__float128, long)
note: built-in candidate operator(__float128, long long)
note: built-in candidate operator(__float128, __int128)
note: built-in candidate operator(__float128, unsigned int)
note: built-in candidate operator(__float128, unsigned long)
note: built-in candidate operator(__float128, unsigned long long)
note: built-in candidate operator(__float128, unsigned __int128)
note: built-in candidate operator(int, float)
note: built-in candidate operator(int, double)
note: built-in candidate operator(int, long double)
note: built-in candidate operator(int, __float128)
note: built-in candidate operator(int, int)
note: built-in candidate operator(int, long)
note: built-in candidate operator(int, long long)
note: built-in candidate operator(int, __int128)
note: built-in candidate operator(int, unsigned int)
note: built-in candidate operator(int, unsigned long)
note: built-in candidate operator(int, unsigned long long)
note: built-in candidate operator(int, unsigned __int128)
note: built-in candidate operator(long, float)
note: built-in candidate operator(long, double)
note: built-in candidate operator(long, long double)
note: built-in candidate operator(long, __float128)
note: built-in candidate operator(long, int)
note: built-in candidate operator(long, long)
note: built-in candidate operator(long, long long)
note: built-in candidate operator(long, __int128)
note: built-in candidate operator(long, unsigned int)
note: built-in candidate operator(long, unsigned long)
note: built-in candidate operator(long, unsigned long long)
note: built-in candidate operator(long, unsigned __int128)
note: built-in candidate operator(long long, float)
note: built-in candidate operator(long long, double)
note: built-in candidate operator(long long, long double)
note: built-in candidate operator(long long, __float128)
note: built-in candidate operator(long long, int)
note: built-in candidate operator(long long, long)
note: built-in candidate operator(long long, long long)
note: built-in candidate operator(long long, __int128)
note: built-in candidate operator(long long, unsigned int)
note: built-in candidate operator(long long, unsigned long)
note: built-in candidate operator(long long, unsigned long long)
note: built-in candidate operator(long long, unsigned __int128)
note: built-in candidate operator(__int128, float)
note: built-in candidate operator(__int128, double)
note: built-in candidate operator(__int128, long double)
note: built-in candidate operator(__int128, __float128)
note: built-in candidate operator(__int128, int)
note: built-in candidate operator(__int128, long)
note: built-in candidate operator(__int128, long long)
note: built-in candidate operator(__int128, __int128)
note: built-in candidate operator(__int128, unsigned int)
note: built-in candidate operator(__int128, unsigned long)
note: built-in candidate operator(__int128, unsigned long long)
note: built-in candidate operator(__int128, unsigned __int128)
note: built-in candidate operator(unsigned int, float)
note: built-in candidate operator(unsigned int, double)
note: built-in candidate operator(unsigned int, long double)
note: built-in candidate operator(unsigned int, __float128)
note: built-in candidate operator(unsigned int, int)
note: built-in candidate operator(unsigned int, long)
note: built-in candidate operator(unsigned int, long long)
note: built-in candidate operator(unsigned int, __int128)
note: built-in candidate operator(unsigned int, unsigned int)
note: built-in candidate operator(unsigned int, unsigned long)
note: built-in candidate operator(unsigned int, unsigned long long)
note: built-in candidate operator(unsigned int, unsigned __int128)
note: built-in candidate operator(unsigned long, float)
note: built-in candidate operator(unsigned long, double)
note: built-in candidate operator(unsigned long, long double)
note: built-in candidate operator(unsigned long, __float128)
note: built-in candidate operator(unsigned long, int)
note: built-in candidate operator(unsigned long, long)
note: built-in candidate operator(unsigned long, long long)
note: built-in candidate operator(unsigned long, __int128)
note: built-in candidate operator(unsigned long, unsigned int)
note: built-in candidate operator(unsigned long, unsigned long)
note: built-in candidate operator(unsigned long, unsigned long long)
note: built-in candidate operator(unsigned long, unsigned __int128)
note: built-in candidate operator(unsigned long long, float)
note: built-in candidate operator(unsigned long long, double)
note: built-in candidate operator(unsigned long long, long double)
note: built-in candidate operator(unsigned long long, __float128)
note: built-in candidate operator(unsigned long long, int)
note: built-in candidate operator(unsigned long long, long)
note: built-in candidate operator(unsigned long long, long long)
note: built-in candidate operator(unsigned long long, __int128)
note: built-in candidate operator(unsigned long long, unsigned int)
note: built-in candidate operator(unsigned long long, unsigned long)
note: built-in candidate operator(unsigned long long, unsigned long long)
note: built-in candidate operator(unsigned long long, unsigned __int128)
note: built-in candidate operator(unsigned __int128, float)
note: built-in candidate operator(unsigned __int128, double)
note: built-in candidate operator(unsigned __int128, long double)
note: built-in candidate operator(unsigned __int128, __float128)
note: built-in candidate operator(unsigned __int128, int)
note: built-in candidate operator(unsigned __int128, long)
note: built-in candidate operator(unsigned __int128, long long)
note: built-in candidate operator(unsigned __int128, __int128)
note: built-in candidate operator(unsigned __int128, unsigned int)
note: built-in candidate operator(unsigned __int128, unsigned long)
note: built-in candidate operator(unsigned __int128, unsigned long long)
note: built-in candidate operator(unsigned __int128, unsigned __int128)
1 warning and 1 error generated.
Compiler returned: 1
```

### PR DESCRIPTION
**Pull request by AaronBallman**
We previously were not adding them to the candidate set and so use of a bit-precise integer as a class member could lead to ambiguous overload sets.

Fixes https://github.com/llvm/llvm-project/issues/82998

### CODE REVIEW

File: clang/lib/Sema/SemaOverload.cpp [hunk 1]
```diff
@@ -8955,6 +8962,22 @@ class BuiltinOperatorOverloadBuilder {
...
+    /// sets, we need to do some extra work to unique things. Copy the
+    /// candidates into a unique set, then copy from that set into the list of
+    /// arithmetic types.
+    llvm::SmallSetVector<CanQualType, 2> BitIntCandidates;
+    llvm::for_each(CandidateTypes, [&BitIntCandidates](
+                                       BuiltinCandidateTypeSet &Candidate) {
+      for (QualType BitTy : Candidate.bitint_types())
+        BitIntCandidates.insert(CanQualType::CreateUnsafe(BitTy));
+    });
+    llvm::copy(BitIntCandidates, std::back_inserter(ArithmeticTypes));
```
Comment (erichkeane) [inline clang/lib/Sema/SemaOverload.cpp:8980]: I'm concerned about the assert on `9001`.  It seems to me that this changes us from being a reasonably 'fixed' list to having a perhaps-unlimited list.  

I think that the intent of that assert is nice (ensuring we can small-vector-opt this list), but this section here makes that no longer guaranteed, right? 

The `ArithmeticTypesCap` is 24, and we do a `push_back` of 23 other types (by my count!).  So adding TWO bit-int types + having all the rest of the types enabled causes that assert to fire.

Comment (AaronBallman) [reply clang/lib/Sema/SemaOverload.cpp]: Done

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/lib/Sema/SemaOverload.cpp]: Done

File: clang/lib/Sema/SemaOverload.cpp [hunk 2]
```diff
@@ -8955,6 +8962,22 @@ class BuiltinOperatorOverloadBuilder {
         (S.Context.getAuxTargetInfo() &&
          S.Context.getAuxTargetInfo()->hasInt128Type()))
       ArithmeticTypes.push_back(S.Context.UnsignedInt128Ty);
+
+    /// _BitInt overloads are a bit special. We don't want to add candidates
+    /// for the entire family of _BitInt types, so instead we only add
+    /// candidates for the unique, unqualified _BitInt types present in the
+    /// candidate type set. The candidate set already handled ensuring the type
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/lib/Sema/SemaOverload.cpp]: Done

**Review by erichkeane — APPROVED** [review]
LGTM if Corentin is happy with the comment.

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks!

**Pull request merged**